File: packages/sdk-python/trustlayer/firewall.py
"""
TrustLayer Python SDK — Firewall Resource
"""

from __future__ import annotations
from typing import List, Optional

from .client import BaseClient
from .types import EvaluateRequest, EvaluateResponse, CheckResult, Verdict, FirewallBlockError


class FirewallResource:
    def __init__(self, client: BaseClient):
        self._client = client

    async def evaluate(
        self,
        request: EvaluateRequest,
        *,
        agent_id: Optional[str] = None,
        throw_on_block: bool = False,
    ) -> EvaluateResponse:
        """
        Evaluates an agent action through the TrustLayer governance pipeline.

        Returns an EvaluateResponse with verdict: 'allow', 'block', or 'escalate'.
        If throw_on_block=True, raises FirewallBlockError on block verdicts.

        Example:
            result = await tl.firewall.evaluate(EvaluateRequest(
                action="delete_user_record",
                resource="crm:contacts",
                description="Delete customer record per GDPR request",
            ))
            if result.verdict == Verdict.ALLOW:
                # proceed
                pass
        """
        body = {
            "action": request.action,
            "resource": request.resource,
        }
        if request.tool_name:
            body["tool_name"] = request.tool_name
        if request.description:
            body["description"] = request.description
        if request.domain:
            body["domain"] = request.domain
        if request.metadata:
            body["metadata"] = request.metadata
        if request.risk_threshold is not None:
            body["risk_threshold"] = request.risk_threshold

        try:
            raw = await self._client.request(
                "POST",
                "/api/v1/firewall/evaluate",
                body=body,
                agent_id=agent_id or request.agent_id,
            )
        except Exception as e:
            # Re-raise — block verdicts come as 403 which BaseClient raises as TrustLayerError
            raise

        result = _map_evaluation(raw)

        if throw_on_block and result.verdict == Verdict.BLOCK:
            raise FirewallBlockError(result)

        return result

    async def list(
        self,
        limit: int = 50,
        verdict: Optional[Verdict] = None,
    ) -> List[EvaluateResponse]:
        """Lists recent firewall evaluations."""
        query = {"limit": str(limit)}
        if verdict:
            query["verdict"] = verdict.value

        raw_list = await self._client.request("GET", "/api/v1/firewall/evaluations", query=query)
        return [_map_evaluation(r) for r in (raw_list or [])]


def _map_evaluation(raw: dict) -> EvaluateResponse:
    return EvaluateResponse(
        evaluation_id=raw.get("evaluation_id") or raw.get("id", ""),
        verdict=Verdict(raw.get("verdict", "allow")),
        risk_score=raw.get("risk_score", 0),
        checks=[
            CheckResult(
                check=c.get("check", ""),
                passed=c.get("passed", True),
                detail=c.get("detail", ""),
                latency_ms=c.get("latency_ms", 0),
            )
            for c in (raw.get("checks") or [])
        ],
        reasons=raw.get("reasons") or [],
        certificate_id=raw.get("certificate_id"),
        hitl_ticket_id=raw.get("hitl_ticket_id"),
        latency_ms=raw.get("latency_ms", 0),
    )

File: packages/sdk-node/src/firewall.ts
/**
 * RuneSignal Node SDK — Firewall Resource
 * POST /api/v1/firewall/evaluate
 */

import { BaseClient } from './client';
import {
  EvaluateRequest,
  EvaluateResponse,
  FirewallBlockError,
} from './types';

export class FirewallResource {
  constructor(private readonly client: BaseClient) {}

  /**
   * Evaluates an agent action through the full RuneSignal governance pipeline.
   *
   * @param request - The action to evaluate
   * @param options - Optional agentId override
   * @returns EvaluateResponse with verdict: 'allow' | 'block' | 'escalate'
   * @throws FirewallBlockError if verdict is 'block'
   *
   * @example
   * const result = await tl.firewall.evaluate({
   *   action: 'delete_user_record',
   *   resource: 'crm:contacts',
   *   description: 'Delete customer record per GDPR request'
   * });
   * if (result.verdict === 'escalate') {
   *   console.log('Waiting for human approval:', result.hitlTicketId);
   * }
   */
  async evaluate(
    request: EvaluateRequest,
    options: { agentId?: string; throwOnBlock?: boolean } = {}
  ): Promise<EvaluateResponse> {
    const { throwOnBlock = false, agentId } = options;

    const body = {
      action: request.action,
      resource: request.resource,
      tool_name: request.toolName,
      description: request.description,
      domain: request.domain,
      metadata: request.metadata,
      risk_threshold: request.riskThreshold,
    };

    // The API returns 403 for block verdicts — we catch that and re-parse
    let raw: any;
    try {
      raw = await (this.client as any).request('POST', '/api/v1/firewall/evaluate', {
        body,
        agentId: agentId || request.agentId,
      });
    } catch (e: any) {
      // 403 from firewall = block verdict, parse the response body
      if (e.status === 403 && e.detail) {
        raw = e.detail;
      } else {
        throw e;
      }
    }

    const result: EvaluateResponse = {
      evaluationId: raw.evaluation_id,
      verdict: raw.verdict,
      riskScore: raw.risk_score,
      checks: (raw.checks || []).map((c: any) => ({
        check: c.check,
        passed: c.passed,
        detail: c.detail,
        latencyMs: c.latency_ms,
      })),
      reasons: raw.reasons || [],
      certificateId: raw.certificate_id,
      hitlTicketId: raw.hitl_ticket_id,
      latencyMs: raw.latency_ms,
    };

    if (throwOnBlock && result.verdict === 'block') {
      throw new FirewallBlockError(result);
    }

    return result;
  }

  /**
   * Lists recent firewall evaluations.
   */
  async list(options: { limit?: number; verdict?: 'allow' | 'block' | 'escalate' } = {}): Promise<EvaluateResponse[]> {
    const query: Record<string, string> = {};
    if (options.limit) query.limit = String(options.limit);
    if (options.verdict) query.verdict = options.verdict;

    const raw: any[] = await (this.client as any).request('GET', '/api/v1/firewall/evaluations', { query });
    return raw.map(r => ({
      evaluationId: r.id,
      verdict: r.verdict,
      riskScore: r.risk_score,
      checks: r.checks || [],
      reasons: r.reasons || [],
      certificateId: r.certificate_id,
      hitlTicketId: r.hitl_ticket_id,
      latencyMs: r.latency_ms,
    }));
  }
}
